src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -693,17 +693,28 @@ public CompletionStats completionStats(String... fields) {
         return completionStats;
     }
 
-    public void flush(FlushRequest request) {
-        // we allows flush while recovering, since we allow for operations to happen
-        // while recovering, and we want to keep the translog at bay (up to deletes, which
-        // we don't gc).
+    public Engine.SyncedFlushResult syncFlushIfNoPendingChanges(String syncId, byte[] expectedCommitId) {
         verifyStartedOrRecovering();
+        logger.trace("trying to sync flush. sync id [{}]. expected commit id [{}]]", syncId, expectedCommitId);
+        return engine().syncFlushIfNoPendingChanges(syncId, expectedCommitId);
+    }
+
+    public byte[] flush(FlushRequest request) throws ElasticsearchException {
```

Comment:
can we add an abstraction to this please? I don't like to return a mutable object as the ID. It can be very simple `CommitID` class that has hashCode & equals etc